File: inception/inception-support/src/main/java/de/tudarmstadt/ukp/inception/support/kendo/AjaxSplitterBehavior.js
/*
 * Licensed to the Technische Universität Darmstadt under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The Technische Universität Darmstadt
 * licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Kendo's splitter parses pane size with parseInt(), truncating fractional percent. To
// preserve the server's stored precision we keep the percent intent here and re-apply
// it as integer pixels (which Kendo handles exactly) on init and on window resize.
function initInceptionAjaxSplitter(selector, url, orientation, pctIntent) {
  const RESIZE_DEBOUNCE_MS = 150;
  const STATE_KEY = 'inceptionAjaxSplitter';
  const vertical = orientation === 'vertical';
  const extentProp = vertical ? 'offsetHeight' : 'offsetWidth';
  const extentFn = vertical ? 'height' : 'width';

  const $splitter = $(selector);
  if (!$splitter.length) return;
  const splitter = $splitter.data('kendoSplitter');
  if (!splitter) return;

  // Wicket may re-render the splitter (e.g. on sidebar toggle), re-running this init. Keep any
  // prior state so a drag survives, but always rebind the handlers below - stale bindings from the
  // previous widget instance silently break size persistence.
  let state = $splitter.data(STATE_KEY);
  if (state) {
    state.url = url;
    if (Array.isArray(pctIntent)) state.pctIntent = pctIntent.slice();
  } else {
    state = {
      url: url,
      // null entries are pixel/auto panes - never persisted as percent. Prefer the percentages
      // passed in by the server: Kendo rewrites each pane's `size` to pixels during init, so
      // reading them back off the widget can capture a mid-layout value and lock the pane there.
      pctIntent: Array.isArray(pctIntent) ? pctIntent.slice()
        : splitter.options.panes.map(paneOption => {
          if (paneOption && typeof paneOption.size === 'string'
              && paneOption.size.indexOf('%') > 0) {
            const parsed = parseFloat(paneOption.size);
            return isNaN(parsed) ? null : parsed;
          }
          return null;
        }),
      // Suppresses the resize handler during programmatic re-layout so rounding drift is never
      // persisted as a user drag.
      applying: false,
      lastSnapshot: null,
      // Extent at which applyPixels() last applied the intent; null = never (it bails on a zero
      // extent). See the resize handler for why null must not count as "extent changed".
      settledExtent: null
    };
    $splitter.data(STATE_KEY, state);
  }

  // Kendo's denominator: element extent minus the splitbars (along the splitter axis).
  const computeAvailableExtent = () => {
    let availableExtent = $splitter[extentFn]();
    if (!availableExtent) return 0;
    $splitter.children('.k-splitbar').each(function () {
      availableExtent -= this[extentProp];
    });
    return availableExtent;
  };

  const applyPixels = () => {
    const availableExtent = computeAvailableExtent();
    if (availableExtent <= 0) return;
    const panes = $splitter.children('.k-pane');
    panes.each(function (i, pane) {
      if (!pane.id) pane.id = $splitter[0].id + '_p' + i;
    });
    state.applying = true;
    try {
      state.pctIntent.forEach((percent, i) => {
        if (percent === null) return;
        const pane = panes.get(i);
        if (!pane) return;
        splitter.size('#' + pane.id, Math.round(percent * availableExtent / 100) + 'px');
      });
      state.lastSnapshot = snapshot();
      state.settledExtent = availableExtent;
    }
    finally {
      state.applying = false;
    }
  };

  const snapshot = () => {
    const availableExtent = computeAvailableExtent();
    if (availableExtent <= 0) return null;
    const sizes = [];
    $splitter.children('.k-pane').each(function (i, pane) {
      sizes.push(pane[extentProp] / availableExtent * 100);
    });
    return sizes.join(',');
  };

  applyPixels();

  // A splitter still being laid out reports a near-zero extent, so the applyPixels() above may
  // have bailed. Re-apply the intent whenever the extent actually changes, covering both the
  // initial settle and later relayouts.
  if (state.sizeObserver) state.sizeObserver.disconnect();
  if (typeof ResizeObserver === 'function') {
    let lastExtent = computeAvailableExtent();
    state.sizeObserver = new ResizeObserver(() => {
      if (state.applying) return;
      const extent = computeAvailableExtent();
      // Sub-pixel jitter must not re-apply, or dragging fights the observer.
      if (extent <= 0 || Math.abs(extent - lastExtent) < 1) return;
      lastExtent = extent;
      applyPixels();
    });
    state.sizeObserver.observe($splitter[0]);
  }

  let resizeTimer;
  const eventNs = '.splitter_' + $splitter[0].id;
  state.eventNs = eventNs;
  $(window).off(eventNs).on('resize' + eventNs, () => {
    clearTimeout(resizeTimer);
    resizeTimer = setTimeout(applyPixels, RESIZE_DEBOUNCE_MS);
  });

  // Unbind only our own handler so listeners registered via SplitterBehavior's
  // ISplitterListener / options.resize survive a re-init.
  if (state.resizeHandler) splitter.unbind('resize', state.resizeHandler);
  state.resizeHandler = () => {
    if (state.applying) return;
    const currentSnapshot = snapshot();
    if (currentSnapshot === null || currentSnapshot === state.lastSnapshot) return;

    // Kendo fires `resize` for any relayout, not just a user drag. Mid-layout ratios must not
    // become the new intent, or the pane is persisted at whatever fraction it held while the
    // container was still settling. So when the extent has changed since we last applied the
    // intent, re-assert it instead of adopting the ratios.
    //
    // Unless we never applied it (settledExtent === null): then there is no known-good intent to
    // restore and re-applying would snap the pane back, discarding this drag. The extent is
    // non-zero here, so the layout has settled and the drag is genuine.
    const availableExtent = computeAvailableExtent();
    if (state.settledExtent !== null && Math.abs(state.settledExtent - availableExtent) >= 1) {
      state.settledExtent = availableExtent;
      applyPixels();
      return;
    }
    state.settledExtent = availableExtent;

    state.lastSnapshot = currentSnapshot;
    $splitter.children('.k-pane').each(function (i, pane) {
      if (state.pctIntent[i] !== null) {
        state.pctIntent[i] = pane[extentProp] / availableExtent * 100;
      }
    });
    Wicket.Ajax.get({ u: state.url, ep: { sizes: currentSnapshot } });
  };
  splitter.bind('resize', state.resizeHandler);
}

// Tear down the Kendo widget and the listeners bound by init. Must run before Wicket replaces the
// splitter DOM, or the lingering widget reference briefly leaves the new panes unstyled.
function destroyInceptionAjaxSplitter(selector) {
  const $splitter = $(selector);
  if (!$splitter.length) return;
  const state = $splitter.data('inceptionAjaxSplitter');
  if (state && state.eventNs) {
    $(window).off(state.eventNs);
  }
  if (state && state.sizeObserver) state.sizeObserver.disconnect();
  $splitter.removeData('inceptionAjaxSplitter');
  const splitter = $splitter.data('kendoSplitter');
  if (splitter) splitter.destroy();
  $splitter.removeData('kendoSplitter');
}

// Re-create the splitter in place with a new pane configuration, reusing the existing pane DOM so
// its content survives (e.g. an editor iframe's scroll position). Use instead of a destroy plus
// full Wicket re-render when pane content must be preserved.
function reconfigureInceptionAjaxSplitter(selector, panes, url, orientation, pctIntent) {
  const $splitter = $(selector);
  if (!$splitter.length) return;
  const state = $splitter.data('inceptionAjaxSplitter');
  if (state && state.eventNs) {
    $(window).off(state.eventNs);
  }
  if (state && state.sizeObserver) state.sizeObserver.disconnect();
  const existing = $splitter.data('kendoSplitter');
  if (existing) existing.destroy();
  $splitter.removeData('kendoSplitter');
  $splitter.removeData('inceptionAjaxSplitter');
  $splitter.children('.k-splitbar').remove();
  $splitter.kendoSplitter({ orientation: orientation, panes: panes });
  initInceptionAjaxSplitter(selector, url, orientation, pctIntent);
}

File: inception/inception-support/src/main/java/de/tudarmstadt/ukp/inception/support/kendo/AjaxSplitterBehavior.java
/*
 * Licensed to the Technische Universität Darmstadt under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The Technische Universität Darmstadt
 * licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.tudarmstadt.ukp.inception.support.kendo;

import static java.lang.Double.parseDouble;
import static java.lang.String.format;
import static org.apache.wicket.markup.head.JavaScriptHeaderItem.forReference;
import static org.wicketstuff.jquery.core.Options.asString;

import org.apache.wicket.Component;
import org.apache.wicket.ajax.AbstractDefaultAjaxBehavior;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.markup.head.IHeaderResponse;
import org.apache.wicket.markup.head.OnDomReadyHeaderItem;
import org.wicketstuff.jquery.core.Options;
import org.wicketstuff.kendo.ui.widget.splitter.ISplitterListener;
import org.wicketstuff.kendo.ui.widget.splitter.SplitterAdapter;
import org.wicketstuff.kendo.ui.widget.splitter.SplitterBehavior;

/**
 * A {@link SplitterBehavior} that additionally reports pane size changes back to the server via an
 * AJAX callback. Subclasses implement {@link #onResize(AjaxRequestTarget, double[])} to react.
 * <p>
 * Sizes are reported as fractional percentages of the splitter's total width (or height), one entry
 * per pane, in DOM order. The callback is invoked only when the size vector actually changes
 * compared to the previous report, so layout re-flows do not generate spurious requests.
 */
public abstract class AjaxSplitterBehavior
    extends SplitterBehavior
{
    private static final long serialVersionUID = 1L;

    public enum Orientation
    {
        HORIZONTAL, VERTICAL;

        String optionValue()
        {
            return name().toLowerCase();
        }
    }

    private final Orientation orientation;
    private AbstractDefaultAjaxBehavior resizeBehavior;
    private Double[] percentIntent;

    public AjaxSplitterBehavior(String aSelector, Orientation aOrientation, Options aOptions)
    {
        this(aSelector, aOrientation, aOptions, new SplitterAdapter());
    }

    public AjaxSplitterBehavior(String aSelector, Orientation aOrientation, Options aOptions,
            ISplitterListener aListener)
    {
        super(aSelector, withOrientation(aOptions, aOrientation), aListener);
        orientation = aOrientation;
    }

    private static Options withOrientation(Options aOptions, Orientation aOrientation)
    {
        var options = aOptions != null ? aOptions : new Options();
        options.set("orientation", asString(aOrientation.optionValue()));
        return options;
    }

    @Override
    public void bind(Component aComponent)
    {
        super.bind(aComponent);

        aComponent.setOutputMarkupId(true);

        resizeBehavior = new AbstractDefaultAjaxBehavior()
        {
            private static final long serialVersionUID = 1L;

            @Override
            protected void respond(AjaxRequestTarget aTarget)
            {
                var raw = aComponent.getRequest().getRequestParameters().getParameterValue("sizes")
                        .toString("");
                if (raw.isEmpty()) {
                    return;
                }

                var parts = raw.split(",");
                var sizes = new double[parts.length];
                try {
                    for (int i = 0; i < parts.length; i++) {
                        sizes[i] = parseDouble(parts[i]);
                    }
                }
                catch (NumberFormatException e) {
                    return;
                }

                onResize(aTarget, sizes);
            }
        };
        aComponent.add(resizeBehavior);
    }

    /**
     * Triggered when the user finishes dragging a splitter pane handle.
     *
     * @param aTarget
     *            the AJAX request target
     * @param aSizesPercent
     *            percentage size of each pane (0-100), in DOM order
     */
    protected abstract void onResize(AjaxRequestTarget aTarget, double[] aSizesPercent);

    /**
     * Schedule destruction of the Kendo splitter widget bound to this behavior. Call this before
     * replacing the splitter's DOM in an AJAX response so the lingering widget reference does not
     * leave the new pane elements briefly unstyled during the swap.
     */
    public void destroy(AjaxRequestTarget aTarget)
    {
        aTarget.prependJavaScript(format("destroyInceptionAjaxSplitter(%s);", asString(selector)));
    }

    /**
     * Re-create the Kendo splitter widget in place with a new pane configuration, without having
     * Wicket re-render the splitter's DOM. Use this instead of {@link #destroy} followed by a full
     * re-render of the splitter container when some pane content must survive the reconfiguration
     * (e.g. an editor IFrame whose scroll position - and the whole editor - would otherwise be
     * lost). Pair it with {@link #destroy}: that call runs as prepended JavaScript and unbinds the
     * old widget while its DOM is still intact; this call re-initialises (as appended JavaScript)
     * after the partial DOM update has been applied.
     *
     * @param aTarget
     *            the Ajax request target
     * @param aPanes
     *            the new pane configuration, one entry per pane in DOM order. The percentage intent
     *            is derived from these, so it cannot disagree with them.
     */
    public void reconfigure(AjaxRequestTarget aTarget, Options... aPanes)
    {
        percentIntent = percentIntentOf(aPanes);

        aTarget.appendJavaScript(format("reconfigureInceptionAjaxSplitter(%s, %s, %s, %s, %s);",
                asString(selector), panesJson(aPanes), asString(resizeBehavior.getCallbackUrl()),
                asString(orientation.optionValue()), percentIntentJson()));
    }

    @Override
    public void renderHead(Component aComponent, IHeaderResponse aResponse)
    {
        super.renderHead(aComponent, aResponse);

        aResponse.render(forReference(AjaxSplitterJavaScriptReference.get()));

        var script = format("initInceptionAjaxSplitter(%s, %s, %s, %s);", asString(selector),
                asString(resizeBehavior.getCallbackUrl()), asString(orientation.optionValue()),
                percentIntentJson());

        aResponse.render(OnDomReadyHeaderItem.forScript(script));
    }

    /**
     * Set the pane widths in in percent.
     *
     * @param aPanes
     *            one entry per pane in DOM order
     */
    protected void setPanes(Options... aPanes)
    {
        setOption("panes", (Object[]) aPanes);
        percentIntent = percentIntentOf(aPanes);
    }

    private static String panesJson(Options... aPanes)
    {
        var json = new StringBuilder("[");
        for (int i = 0; i < aPanes.length; i++) {
            if (i > 0) {
                json.append(",");
            }
            json.append(aPanes[i]);
        }
        return json.append("]").toString();
    }

    private static Double[] percentIntentOf(Options... aPanes)
    {
        var percentages = new Double[aPanes.length];
        for (int i = 0; i < aPanes.length; i++) {
            percentages[i] = percentOrNull(aPanes[i]);
        }
        return percentages;
    }

    private static Double percentOrNull(Options aPane)
    {
        var size = aPane.get("size");
        if (size == null) {
            return null;
        }

        // Options renders string values quoted, so strip those along with the percent sign.
        var text = size.toString().replace("\"", "").trim();
        if (!text.endsWith("%")) {
            return null;
        }

        try {
            return Double.valueOf(text.substring(0, text.length() - 1));
        }
        catch (NumberFormatException e) {
            return null;
        }
    }

    private String percentIntentJson()
    {
        if (percentIntent == null) {
            return "null";
        }

        var json = new StringBuilder("[");
        for (int i = 0; i < percentIntent.length; i++) {
            if (i > 0) {
                json.append(",");
            }
            json.append(percentIntent[i] == null ? "null" : percentIntent[i].toString());
        }
        return json.append("]").toString();
    }
}
